app/jobs/base.rb

```diff
@@ -16,35 +16,87 @@ def self.num_email_retry_jobs
+        @data["job_type"] = job_class.try(:scheduled?) ? "scheduled" : "regular" # Job Type - either s for scheduled or r for regular
+        @data["opts"] = opts # Params - json encoded params for the job
+
+        @data["status"] = 'pending'
+        @start_timestamp = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+        self.class.ensure_interval_logging!
+        @@active_jobs ||= []
+        @@active_jobs << self
+
+        MethodProfiler.ensure_discourse_instrumentation!
+        MethodProfiler.start
+      end
 
-    class Instrumenter
+      def stop(exception:)
+        return unless enabled?
 
-      def self.stats
-        Thread.current[:db_stats] ||= Stats.new
-      end
+        profile = MethodProfiler.stop
+
+        @@active_jobs.delete(self)
 
-      class Stats
-        attr_accessor :query_count, :duration_ms
+        @data["duration"] = profile[:total_duration] # Duration - length in seconds it took to run
+        @data["sql_duration"] = profile.dig(:sql, :duration) || 0 # Sql Duration (s)
+        @data["sql_calls"] = profile.dig(:sql, :calls) || 0 # Sql Statements - how many statements ran
+        @data["redis_duration"] = profile.dig(:redis, :duration) || 0 # Redis Duration (s)
+        @data["redis_calls"] = profile.dig(:redis, :calls) || 0 # Redis commands
+        @data["net_duration"] = profile.dig(:net, :duration) || 0 # Redis Duration (s)
+        @data["net_calls"] = profile.dig(:net, :calls) || 0 # Redis commands
 
-        def initialize
-          @query_count = 0
-          @duration_ms = 0
+        if exception.present?
+          @data["exception"] = exception # Exception - if job fails a json encoded exception
+          @data["status"] = 'failed'
+        else
+          @data["status"] = 'success' # Status - fail, success, pending
         end
+
+        write_to_log
       end
 
-      def call(name, start, finish, message_id, values)
-        stats = Instrumenter.stats
-        stats.query_count += 1
-        stats.duration_ms += (((finish - start).to_f) * 1000).to_i
+      def logger
+        @@logger ||= Logger.new("#{Rails.root}/log/sidekiq.log")
       end
-    end
 
-    include Sidekiq::Worker
+      def current_duration
+        Process.clock_gettime(Process::CLOCK_MONOTONIC) - @start_timestamp
+      end
+
+      def write_to_log
+        return unless enabled?
+        @data["@timestamp"] = Time.now
+        @data["duration"] = current_duration if @data["status"] == "pending"
+        logger << "#{@data.to_json}\n"
+      end
 
-    def initialize
-      @db_duration = 0
+      def enabled?
+        ENV["DISCOURSE_LOG_SIDEKIQ"] == "1"
+      end
+
+      def self.ensure_interval_logging!
+        interval = ENV["DISCOURSE_LOG_SIDEKIQ_INTERVAL"]
+        return if !interval
+        @@interval_thread ||= Thread.new do
+          loop do
+            sleep interval.to_i
+            @@active_jobs.each { |j| j.write_to_log if j.current_duration > interval }
```

Comment:
got to output something somewhere if this thread crashes.

Reply:
Done in f9cf37b